test/test_calendar.py

```diff
@@ -150,3 +156,107 @@ def test_parse_date_offset_no_offset_magnitude():
+                datetime.datetime(year=301, month=1, day=1),
+                GREGORIAN,
+                365 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=400, month=1, day=1),
+                datetime.datetime(year=401, month=12, day=31),
+                GREGORIAN,
+                365 * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=12, month=7, day=22),
+                datetime.datetime(year=23, month=3, day=15),
+                GREGORIAN,
+                (10 * 365 + 238) * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=1, month=1, day=1),
+                datetime.datetime(year=1001, month=1, day=1),
+                GREGORIAN,
+                (1000 * 365 + 242) * SEC_PER_DAY
+            ),
+            (
+                datetime.datetime(year=1, month=1, day=1),
+                datetime.datetime(year=9999, month=1, day=1),
+                GREGORIAN,
+                (9998 * 365 + 2424) * SEC_PER_DAY
+            )
+        ]
+)
+def test_seconds_between_dates(start_date, end_date, caltype_int, expected):
+    assert seconds_between_dates(start_date, end_date, caltype_int) == expected
+
+
+@pytest.mark.parametrize(
+        "date_int, expected",
+        [
+            (10101, datetime.date(1, 1, 1)),
+            (100321, datetime.date(10, 3, 21)),
+            (38211130, datetime.date(3821, 11, 30))
+        ]
+)
+def test_int_to_date(date_int, expected):
+    """
+    Check that integers typically read in from namelists
+    are correctly converted to datetime.date objects.
+    """
+    converted_date = int_to_date(date_int)
+    assert converted_date == expected
+
+
+@pytest.mark.parametrize(
+        "start_date, years, months, days, seconds, caltype, expected",
+        [
+            (datetime.date(101, 1, 1), 1, 0, 0, 0, GREGORIAN, 365*SEC_PER_DAY),
+            (datetime.date(40, 2, 8), 0, 1, 0, 0, GREGORIAN, 29*SEC_PER_DAY),
+            (datetime.date(40, 2, 8), 0, 1, 0, 0, NOLEAP, 28*SEC_PER_DAY),
+            (datetime.date(153, 8, 21), 2, 7, 510, 5321,
+             GREGORIAN, 1453*SEC_PER_DAY + 5321),
+            # Extreme cases, unlikely to ever happen
+            (datetime.date(1, 1, 1), 9998, 0, 0, 0,
+             NOLEAP, 9998 * 365 * SEC_PER_DAY),
+            (datetime.date(1, 1, 1), 9998, 0, 0, 0,
+             GREGORIAN, (9998 * 365 + 2424) * SEC_PER_DAY)
+        ]
+)
+def test_runtime_from_date(
+        start_date,
+        years,
+        months,
+        days,
+        seconds,
+        caltype,
+        expected):
+    """
+    Test that the number of seconds calculated for run lengths is correct.
+    """
+    runtime = runtime_from_date(start_date,
+                                years,
+                                months,
+                                days,
+                                seconds,
+                                caltype)
+
+    assert runtime == expected
```

Comment:
Again - we should test the negative cases ... year -1, month 13, etc

Reply:
This is a good point – It turns out that this function currently doesn't guard against bad runtimes being specified in the `config.yaml`. If it's given a negative input it outputs a negative number of seconds for the runtime.

Since this is something that touches a couple other models, do you think it would be ok to defer adding guard rails to a separate issue?

Reply:
Ok i raised https://github.com/payu-org/payu/issues/492